### ISSUE TITLE
Base64 decode with altchars still accepts non-altchars

### ISSUE THREAD

**Issue by Anonymous3-a** [python/cpython#125346]
# Bug report

### Bug description:

If you run
```python
>>> import base64
>>> base64.b64decode("/+/+")
```
it gives
```python
b'\xff\xef\xfe'
```
This is expected.
If you run
```python
>>> base64.b64decode("_-_-", altchars="-_")
```
It gives the same output.
This is expected.
However, if you run
```python
>>> base64.b64decode("/+/+", altchars="-_")
```
It also gives the same output.
This is not expected.

This is concerning because I'm trying to use file safe base64 (urlsafe_b64decode/encode do not have validate arg, which I need). I can work around this by also checking for / and + as well.

### CPython versions tested on:

3.9, 3.11

### Operating systems tested on:

Linux, Other

### Linked PRs
* gh-141061
* gh-141130
* gh-141133
* gh-141128
* gh-141153

**Comment by ImFeH2** [comment]
It's intentional behavior that was decided upon years ago. found the following as i'm trying to fix, which related to a commit in 2016.
https://github.com/python/cpython/blob/4a943c3251d1b3fdf50cfb9264ae74e5bc845c3c/Lib/test/test_base64.py#L266-L269

But I think the original alphabet should be discarded because an alphabet should always remain singular. if alternative characters are provided, the original ones should be deprecated to avoid confusion and maintain consistency.

**Comment by picnixz** [comment]
Should we document this behaviour? cc @serhiy-storchaka

**Comment by serhiy-storchaka** [comment]
We should look at the discussion when that feature was added, but I think it is not intentional behavior.

**Comment by Anonymous3-a** [comment]
>I think it is not intentional behavior

See https://github.com/python/cpython/issues/125346#issuecomment-2408454116

**Comment by vadmium** [comment]
My test only tests with validate=False, and I only added it to ensure my documentation fixes matched the existing behaviour; see <https://bugs.python.org/issue22088#msg256368>.

Looking further back, _validate_ was added in Git commit 6495136e4018. The _altchars_ aspect doesn't seem to have been discussed in the associated bug thread <https://bugs.python.org/issue1466065>, so it's not clear if it was considered or not.

**Comment by Hans-Peter-Cad** [comment]
Perhaps the priority or severity of this ticket should be reconsidered ...

What's the sense of parameter 'validate' ?

As I tried to say in my (duplicate) ticket, I consider the current behaviour of 'b64decode' as a (severe) security hole. Inside an encoded string or file, I would expect that **any manipulation** of it would be detected during decoding - if 'validate' is set.

(Here, seeing a message like "binascii.Error: Only base64 data is allowed", also for non-altchars.)

But that's not so in the current implementation (up to 3.13.7); thus giving the possibility to patch an encoded file and to smuggle further information inside it.

**Comment by serhiy-storchaka** [comment]
This behavior was from the beginning (4c904d1bf71d01bc22fbb123493f975050560e9c), but I think it was not intentional. cc @warsaw

For security considerations (this is not an immediate threat, just for general strengthening) we should forbid characters `+` and `/` if `altchars` is used (and it does not contain such characters). But this is a breaking change. Some code can use `urlsafe_b64decode()` and `b64decode()` with `altchars` for data with standard alphabet, or for case when the data can use standard  or alternate alphabets. Changing the behavior will make the working code failing or producing different (maybe more correct!) result.

There is something similar situation with the map01 argument in `b32decode()`, but in that case I think that the behavior is correct -- `0` and `1` should be recognized _in addition_ to `O` and `L` or `I`.

**Comment by Anonymous3-a** [comment]
I feel like not many people are going to have their programs broken by forbidding something that is not useful nor likely used. On the other hand, I am not a Python developer--- I only use it for programming. So, do what you will. Just my 2¢.

Edit: Also, from what I read, people will likely have assumed it scraps the altchars anyways.

**Comment by serhiy-storchaka** [comment]
Bases on the private discussion, accepting "+" and "/" with alternative alphabet only was deprecated.

The user can check that "+" and "/" do not occur in the input before passing it to the decoding function.

**Comment by serhiy-storchaka** [comment]
No, it will not be backported, because there is a risk of breaking existing "working" code, when the code uses `altchars`, but the input actually uses the standard alphabet. This is a programming error, but the behavior should not be changed so drastically in a bugfix release, even a warning is too much.

User can check that neither "+" nor "/" occurr in the input before decoding. They need to do this in 3.15 to avoid warnings (or they can ignore warnings).

In 3.15+ you can also pass the `ignorechars` argument (even `ignorechars=b''` works) to enable more strict and explicit behavior. If "+" and "/" in ignorechars, they will be silently ignored, otherwise they will be errors.
